---
title: "3_1 Transcriptomic data analysis"
author: "Manuel Jara-Espejo"
output: html_notebook
---

### 1.Introduction
After mapping to *Phaw5.1* assembly, embryo RNA-seq data was quantified and analysed

```{r}
knitr::opts_chunk$set(message=F,warning=F,echo=T)
#load libraries
library(tidyr)
library(dplyr)
library(ggplot2)
library(reshape2)
library(ggpubr)
library(colortools)
library(Rsubread)
library(tibble)
```

### 2 Count reads - Embryo PolyA RNA-seq libraries

```{r}
#Call PolyA RNA-seq libraries
# 0-9.15hs, 11-12cells and 24-60hs samples
s1s9_bams_dir <- "/drives/raid/AboobakerLab/data/bams/phaw/hisat/bams_05_2022/Aboobaker_embryo_data/polyA" 
s1s9_bamFiles <- grep("S[1-9].bam$",list.files(s1s9_bams_dir,full.names = T),value=TRUE)

# 16, 20, 24 and 48 hs samples
p16p48_bams_dir <- "/drives/raid/AboobakerLab/data/bams/phaw/hisat/bams_05_2022/Aboobaker_embryo_data/polyA" 
p16p48_bamFiles <- grep("P[0-9]+.bam$",list.files(p16p48_bams_dir,full.names = T),value=TRUE)

#12 and 14 hs samples
p12p14_bams_dir <- "/drives/raid/AboobakerLab/data/bams/phaw/hisat/bams_polya_082022" 
p12p14_bamFiles <- grep('P[0-9]+.bam$',list.files(p12p14_bams_dir,full.names = T),value=TRUE) 

embryo_bamFiles_polyA <- c(s1s9_bamFiles,p16p48_bamFiles,p12p14_bamFiles)
embryo_bamFiles_polyA %>% head(5)
```

```{r}
embryo_geneCounts_polyA <- featureCounts(embryo_bamFiles_polyA,
                                  annot.ext =  "/drives/ssd1/manuel/phaw/2022_analysis/annotation/protein_coding_genes/transdecoder/unigenes/transcripts_cdhit_0.95.fa.transdecoder.genome_unigenes.gtf",
                                  isGTFAnnotationFile= TRUE, GTF.featureType = "transcript",GTF.attrType = "gene_id",
                                  useMetaFeatures= TRUE, allowMultiOverlap = FALSE,minOverlap = 2,
                                  strandSpecific = 2, isPairedEnd= TRUE,nthreads = 8)

embryo_exonCounts_polyA <- featureCounts(embryo_bamFiles_polyA,
                                         annot.ext =  "/drives/ssd1/manuel/phaw/2022_analysis/annotation/protein_coding_genes/transdecoder/unigenes/transcripts_cdhit_0.95.fa.transdecoder.genome_unigenes.gtf",
                                         isGTFAnnotationFile= TRUE,GTF.featureType = "exon",GTF.attrType = "gene_id",
                                         useMetaFeatures= TRUE, allowMultiOverlap = FALSE,minOverlap = 2,
                                         strandSpecific = 2, isPairedEnd= TRUE,nthreads = 8)

#Calculate intron counts by subtracting exon from gene
embryo_intronCounts_polyA = embryo_geneCounts_polyA$counts - embryo_exonCounts_polyA$counts
embryo_intronCounts_polyA=pmax(embryo_intronCounts_polyA,0)
embryo_intronCounts_polyA = as.data.frame(embryo_intronCounts_polyA)
```

```{r}
embryo_exonCounts_polyA$counts %>% head(5)
```

### 3 Count reads - Embryo Total RNA-seq libraries


```{r}
t1t9_bams_dir <- "/drives/raid/AboobakerLab/data/bams/phaw/hisat/bams_05_2022/Aboobaker_embryo_data/total/" # 0-9.15hs, 11-12cells and 24-60hs samples
t1t9_bamFiles <- grep("T[1-9].bam$",list.files(t1t9_bams_dir,full.names = T),value=TRUE) 

t16t48_bams_dir <- "/drives/raid/AboobakerLab/data/bams/phaw/hisat/bams_05_2022/Aboobaker_embryo_data/total/" # 16, 20, 24 and 48 hs samples
t16t48_bamFiles <- grep("T[0-9]{3}.bam$",list.files(t16t48_bams_dir,full.names = T),value=TRUE) 

t12t14_bams_dir <- "/drives/raid/AboobakerLab/data/bams/phaw/hisat/bams_total_082022/" #12 and 14 hs samples
t12t14_bamFiles <- grep('.bam$',list.files(t12t14_bams_dir,full.names = T),value=TRUE) 

embryo_bamFiles_total <- c(t1t9_bamFiles,t16t48_bamFiles,t12t14_bamFiles)

embryo_geneCounts_total <- featureCounts(embryo_bamFiles_total,
                                         annot.ext =  "/drives/ssd1/manuel/phaw/2022_analysis/annotation/protein_coding_genes/transdecoder/unigenes/transcripts_cdhit_0.95.fa.transdecoder.genome_unigenes.gtf",
                                         isGTFAnnotationFile= TRUE,
                                         GTF.featureType = "transcript",
                                         GTF.attrType = "gene_id",
                                         useMetaFeatures= TRUE,
                                         allowMultiOverlap = FALSE,
                                         minOverlap = 2,
                                         strandSpecific = 2, #CHECK THIS PARAMETER
                                         isPairedEnd= TRUE,
                                         nthreads = 8)

embryo_exonCounts_total <- featureCounts(embryo_bamFiles_total,
                                         annot.ext =  "/drives/ssd1/manuel/phaw/2022_analysis/annotation/protein_coding_genes/transdecoder/unigenes/transcripts_cdhit_0.95.fa.transdecoder.genome_unigenes.gtf",
                                         isGTFAnnotationFile= TRUE,
                                         GTF.featureType = "exon",
                                         GTF.attrType = "gene_id",
                                         useMetaFeatures= TRUE,
                                         allowMultiOverlap = FALSE,
                                         minOverlap = 2,
                                         strandSpecific = 2, #CHECK THIS PARAMETER
                                         isPairedEnd= TRUE,
                                         nthreads = 8)

#Calculate intron counts by subtracting exon from gene
embryo_intronCounts_total = embryo_geneCounts_total$counts - embryo_exonCounts_total$counts
embryo_intronCounts_total=pmax(embryo_intronCounts_total,0)
embryo_intronCounts_total = as.data.frame(embryo_intronCounts_total)
```


```{r}
embryo_exonCounts_total$counts %>% head(5)
```

### 4 Calculate TPM values

```{r}
#Open intron and exon total length by gene
exon_sum_byGene <- read.table("/drives/ssd1/manuel/phaw/2022_analysis/annotation/protein_coding_genes/exon_length_byGene.txt",header = T)
intron_sum_byGene <- read.table("/drives/ssd1/manuel/phaw/2022_analysis/annotation/protein_coding_genes/intron_length_byGene.txt",header = T)

#Create function to calculate TPM
calculate_tpm <- function(count_data,length_data) {
  
counts <- left_join(length_data %>% dplyr::select(-3) %>% rename(Length = 2),
                    as_tibble(count_data) %>% mutate(gene_id = rownames(count_data)),
                    by = 'gene_id') %>% na.omit %>% as_tibble() 
  
  #Get mapped reads to transcript normalized by transcript length in Kb
  rate <- counts[,3:ncol(counts)]/(counts$Length/1000)
  counts_tpm <- cbind(gene_id = counts$gene_id, rate) %>% as_tibble()

  #Loop to get the final TPM counts table
  for(i in 1:(dim(counts_tpm)[2]-1)){
    counts_tpm[,i+1] <- counts_tpm[,i+1]/sum(counts_tpm[,i+1])*1e6
  }
  return(counts_tpm)

}

embryo_exonCounts_polyA_tpm <- calculate_tpm(embryo_exonCounts_polyA$counts,exon_sum_byGene)
embryo_exonCounts_total_tpm <- calculate_tpm(embryo_exonCounts_total$counts,exon_sum_byGene)
embryo_intronCounts_polyA_tpm <- calculate_tpm(embryo_intronCounts_masked_polyA,intron_sum_byGene)
colnames(embryo_intronCounts_polyA_tpm) = gsub(".masked","",colnames(embryo_intronCounts_polyA_tpm))
embryo_intronCounts_total_tpm <- calculate_tpm(embryo_intronCounts_masked_total,intron_sum_byGene)
colnames(embryo_intronCounts_total_tpm) = gsub(".masked","",colnames(embryo_intronCounts_total_tpm))

write.table(embryo_exonCounts_polyA_tpm,"embryo_exonCounts_polyA_tpm_byLib.txt",sep = "\t",quote = F,row.names = T )
write.table(embryo_exonCounts_total_tpm,"embryo_exonCounts_total_tpm_byLib.txt",sep = "\t",quote = F,row.names = T )

#Check Nptebook 3_2 to get details about masking process
```

### 5 Visualise RNA-seq data features
Principal Component Analysis (PCA)

```{r}
#Load libraries
library("DESeq2")
library(RColorBrewer)
library(corrplot)
```


```{r}
#PCA Aboobaker data - POLYA-RNAseq
#Create sample files table
sampleFiles <- c("P161.bam","P162.bam","P201.bam", "P202.bam", "P241.bam", "P242.bam", "S2.bam", "S8.bam",
                 "S1.bam","S4.bam","S7.bam","S3.bam","S6.bam","S9.bam","P121.bam","P122.bam","P123.bam",
                 "P141.bam","P142.bam","P143.bam")
stages <- factor(c(rep(c('16hs','20hs','24hs','11hs'),each=2),rep(c('0-9hs','24-60hs',"12hs","13-14hs"),each=3)),levels = c('0-9hs','11hs','12hs','13-14hs','16hs','20hs','24hs','24-60hs'))

sampleTable <- data.frame(sampleName = sampleFiles,
                          condition = stages)
sampleTable
```

```{r}
#Load count matrix into DeSEQ and create PCA plot
dds <- DESeqDataSetFromMatrix(countData = embryo_exonCounts_polyA$counts[,sampleFiles], 
                              colData = sampleTable,
                              design = ~ condition)

pc_embryoPolya <- dds %>% vst %>% plotPCA(ntop=1000,returnData = TRUE)
percentVar <- round(100 * attr(pc_embryoPolya, "percentVar"))
pc_embryoPolya <- ggplot(pc_embryoPolya, aes(PC1, PC2, color=condition)) +
  geom_point(size=3) + labs(color = "Developmental time") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + annotate("text", x=55, y=55, label= "PolyA RNA",size=5)
```

```{r}
#PCA Aboobaker data - Total-RNAseq
sampleFiles <- c("T161.bam","T162.bam","T201.bam", "T202.bam", "T241.bam", "T242.bam","T2.bam","T8.bam",
                 "T1.bam","T4.bam","T7.bam", "T3.bam", "T6.bam","T9.bam","T121.Mixed.bam","T122.Mixed.bam",
                 "T123.Mixed.bam","T141.Mixed.bam","T142.Mixed.bam","T143.Mixed.bam")
stages <- factor(c(rep(c('16hs','20hs','24hs','11hs'),each=2),rep(c('0-9hs','24-60hs',"12hs","13-14hs"),each=3)),levels = c('0-9hs','11hs','12hs','13-14hs','16hs','20hs','24hs','24-60hs'))

sampleTable <- data.frame(sampleName = sampleFiles,
                          condition = stages)
sampleTable
```

```{r}
#Create PCA plot
dds <- DESeqDataSetFromMatrix(countData = embryo_exonCounts_total$counts[,sampleFiles], 
                              colData = sampleTable,
                              design = ~ condition)
pc_embryoTotal <- dds %>% vst %>% plotPCA(ntop=1000,returnData = TRUE)
percentVar <- round(100 * attr(pc_embryoTotal, "percentVar"))
pc_embryoTotal <- ggplot(pc_embryoTotal, aes(PC1, PC2, color=condition)) +
  geom_point(size=3) + labs(color = "Developmental time") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + annotate("text", x=55, y=55, label= "Total RNA",size=5)
```

```{r}
pc_bothData = ggarrange(pc_embryoPolya,pc_embryoTotal,nrow=1,ncol=2,common.legend = TRUE,legend = "bottom")
```


```{r}
#tiff("pca_embryoRNAseq.tiff", width = 10, height = 5.5, units = "in",res = 300,compression = "lzw")
pc_bothData
#dev.off()
```
### 6 Exonic correlations by library

```{r}
#PolyA RNA-seq counts
#Reorder exon counts columns to place stage-specific replicates together
exon_AA_correldata <-  embryo_exonCounts_polyA$counts[,c(1,4,7,2,8,18:20,21:23,10,11,12,13,14,15,3,6,9)] # PolyA
colnames(exon_AA_correldata)= c(rep(c('0-9',"11","12","13-14"),times=c(3,2,3,3)),rep(c('16','20','24','24-60'),times=c(2,2,2,3)))
exonic_cordata_byLibs <- cor(log2(exon_AA_correldata +0.01),use = "complete.obs",method = "pearson")
col <- colorRampPalette(c("lightblue1","lightblue3","lightblue4"))(30)
#Plot correlation coefficients
corrplot(exonic_cordata_byLibs,method = "color",cl.lim = c(0,1),col= col,type = "upper",addCoef.col = "black",tl.cex = 0.8,is.corr=FALSE,number.cex = 0.5)
```

```{r}
#Total RNA-seq counts
exon_AA_correldata <-  embryo_exonCounts_total$counts[,c(1,4,7,2,8,18:20,21:23,10,11,12,13,14,15,3,6,9)] # total
colnames(exon_AA_correldata)= c(rep(c('0-9',"11","12","13-14"),times=c(3,2,3,3)),rep(c('16','20','24','24-60'),times=c(2,2,2,3)))
exonic_cordata_byLibs <- cor(log2(exon_AA_correldata +0.01),use = "complete.obs",method = "pearson")
#Plot correlation coefficients
corrplot(exonic_cordata_byLibs,method = "color",cl.lim = c(0,1),col= col,type = "upper",addCoef.col = "black",tl.cex = 0.8,is.corr=FALSE,number.cex = 0.5)
```
### 7 Plot stage-specific correlations between PolyA and Total TPM counts, for exon and intron data

```{r}
#Summarise TPM counts by stage for each library type
embryo_exon_polya_tpm <- embryo_exonCounts_polyA_tpm %>% as.data.frame() %>% rowwise() %>%
  mutate(p3.5 = mean(c(S1.bam,S4.bam,S7.bam)),
         p4.4 = mean(c(S2.bam,S8.bam)), #Remove second replicate S5.bam
         p4.8 = mean(c(P121.bam,P122.bam,P123.bam)),
         p5.5 = mean(c(P141.bam,P142.bam,P143.bam)),
         p6.5 = mean(c(P161.bam,P162.bam)),
         p8 = mean(c(P201.bam,P202.bam)),
         p10 = mean(c(P241.bam,P242.bam)),
         p19 = mean(c(P481.bam,P482.bam)),
         p10_24= mean(c(S3.bam,S6.bam,S9.bam))) %>% #Stage names based on % of developmental progress. Was replaced later by hours
  dplyr::select(gene_id,p3.5,p4.4,p4.8,p5.5,p6.5,p8,p10,p10_24) #Remove stage 48hrs, both replicates are too discordant

embryo_exon_total_tpm <- embryo_exonCounts_total_tpm %>% as.data.frame() %>% rowwise() %>%
  mutate(p3.5 = mean(c(T1.bam,T4.bam,T7.bam)),
         p4.4 = mean(c(T2.bam,T8.bam)),
         p4.8 = mean(c(T121.Mixed.bam,T122.Mixed.bam,T123.Mixed.bam)),
         p5.5 = mean(c(T141.Mixed.bam,T142.Mixed.bam,T143.Mixed.bam)),
         p6.5 = mean(c(T161.bam,T162.bam)),
         p8 = mean(c(T201.bam,T202.bam)),
         p10 = mean(c(T241.bam,T242.bam)),
         p19 = mean(c(T481.bam,T482.bam)),
         p10_24= mean(c(T3.bam,T6.bam,T9.bam))) %>% #Stage names based on % of developmental progress%>%
  dplyr::select(gene_id,p3.5,p4.4,p4.8,p5.5,p6.5,p8,p10,p10_24) 

embryo_intron_polya_tpm <- embryo_intronCounts_polyA_tpm %>% as.data.frame() %>% rowwise() %>%
  mutate(p3.5 = mean(c(S1.bam,S4.bam,S7.bam)),
         p4.4 = mean(c(S2.bam,S8.bam)),
         p4.8 = mean(c(P121.bam,P122.bam,P123.bam)),
         p5.5 = mean(c(P141.bam,P142.bam,P143.bam)),
         p6.5 = mean(c(P161.bam,P162.bam)),
         p8 = mean(c(P201.bam,P202.bam)),
         p10 = mean(c(P241.bam,P242.bam)),
         p19 = mean(c(P481.bam,P482.bam)),
         p10_24= mean(c(S3.bam,S6.bam,S9.bam))) %>% #Stage names based on % of developmental progress%>%
  dplyr::select(gene_id,p3.5,p4.4,p4.8,p5.5,p6.5,p8,p10,p10_24) 

embryo_intron_total_tpm <- embryo_intronCounts_total_tpm %>% as.data.frame() %>% rowwise() %>%
  mutate(p3.5 = mean(c(T1.bam,T4.bam,T7.bam)),
         p4.4 = mean(c(T2.bam,T8.bam)),
         p4.8 = mean(c(T121.Mixed.bam,T122.Mixed.bam,T123.Mixed.bam)),
         p5.5 = mean(c(T141.Mixed.bam,T142.Mixed.bam,T143.Mixed.bam)),
         p6.5 = mean(c(T161.bam,T162.bam)),
         p8 = mean(c(T201.bam,T202.bam)),
         p10 = mean(c(T241.bam,T242.bam)),
         p19 = mean(c(T481.bam,T482.bam)),
         p10_24= mean(c(T3.bam,T6.bam,T9.bam))) %>% #Stage names based on % of developmental progress%>%
  dplyr::select(gene_id,p3.5,p4.4,p4.8,p5.5,p6.5,p8,p10,p10_24) 
```

#### 7.1 Plot stage-specific correlations between PolyA and Total RNA EXON TPM counts

```{r}
for (i in 1:8) {
  plot_ids <- c("p1","p2", "p3","p4","p5", "p6","p7","p8")
  axis_titles = c('0-9','11','12','13-14','16','20','24','24-60') 
  dataset_list <- list(embryo_exon_polya_tpm,embryo_exon_total_tpm,embryo_intron_polya_tpm,embryo_intron_total_tpm)

    output_plot <- merge(embryo_exon_polya_tpm,embryo_exon_total_tpm,by="gene_id") %>%
    dplyr::rename(polyAcol = i + 1, totalcol = i + 9 ) %>% select(polyAcol,totalcol) %>% 
    ggplot(aes(x= log2(polyAcol +1) ,y= log2(totalcol + 1))) + geom_pointdensity(size=0.2,shape=20) + 
    xlab(NULL) + stat_cor(method = "spearman",size=4) + 
    xlab(paste0(axis_titles[i], " hrs TPM (log2)")) + ylab(paste0(axis_titles[i], " hrs TPM (log2)"))  + stat_cor(method = "spearman",size=4) +
    theme(axis.title = element_text(size = 18),legend.position = "none",axis.text = element_text(size = 14),
          plot.title = element_text(size=20, hjust = 0.5)) + theme_bw()

  assign(plot_ids[i], output_plot)
  rm(output_plot)
}  
exonCorrel_byLibType <- ggarrange(p1,p2,p3,p4,p5,p6,p7,p8,nrow=2,ncol=4, common.legend = TRUE, legend="none")
exonCorrel_byLibType <- annotate_figure(exonCorrel_byLibType, left = text_grob("Total RNA-seq", face = "bold", size = 18,rot = 90,color = "red"),
                bottom = text_grob("PolyA RNA-seq", face = "bold", size = 18,rot = 0,color="blue"))

rm(p1,p2,p3,p4,p5,p6,p7,p8)
```

```{r fig1, fig.height = 5, fig.width = 9}
#tiff("exonCorrel_byLibType.tiff", width = 12, height = 5.5, units = "in",res = 300,compression = "lzw")
exonCorrel_byLibType
#dev.off()
```
#### 7.2 Plot stage-specific correlations between PolyA and Total RNA INTRON TPM counts

```{r}
for (i in 1:8) {
  plot_ids <- c("p1","p2", "p3","p4","p5", "p6","p7","p8")
  axis_titles = c('0-9','11','12','13-14','16','20','24','24-60') 

  output_plot <- merge(embryo_intron_polya_tpm,embryo_intron_total_tpm,by="gene_id") %>%
    dplyr::rename(polyAcol = i + 1, totalcol = i + 9 ) %>% select(polyAcol,totalcol) %>% 
    ggplot(aes(x= log2(polyAcol +1) ,y= log2(totalcol + 1))) + geom_pointdensity(size=0.2,shape=20) + 
    xlab(NULL) + stat_cor(method = "spearman",size=4) + 
    xlab(paste0(axis_titles[i], " hrs TPM (log2)")) + ylab(paste0(axis_titles[i], " hrs TPM (log2)"))  + stat_cor(method = "spearman",size=4) +
    theme(axis.title = element_text(size = 18),legend.position = "none", axis.text = element_text(size = 14),
          plot.title = element_text(size=20, hjust = 0.5)) + theme_bw()
  
  assign(plot_ids[i], output_plot)
  rm(output_plot)
}  
intronCorrel_byLibType <- ggarrange(p1,p2,p3,p4,p5,p6,p7,p8,nrow=2,ncol=4, common.legend = TRUE, legend="none")
intronCorrel_byLibType <- annotate_figure(intronCorrel_byLibType, left = text_grob("Total RNA-seq", face = "bold", size = 18,rot = 90,color = "red"),
                                        bottom = text_grob("PolyA RNA-seq", face = "bold", size = 18,rot = 0,color="blue"))
rm(p1,p2,p3,p4,p5,p6,p7,p8)
```

```{r fig2, fig.height = 5, fig.width = 9}
#tiff("intronCorrel_byLibType.tiff", width = 12, height = 5.5, units = "in",res = 300,compression = "lzw")
intronCorrel_byLibType
#dev.off()
```
## 8 Stage-specific expression boxcplots


```{r}
for (i in 1:4) {
  plot_titles = c("Exon PolyA", "Exon Total", "Intron PolyA","Intron Total")
  plot_ids <- c("p1","p2", "p3","p4")
  dataset_list <- list(embryo_exon_polya_tpm,embryo_exon_total_tpm,embryo_intron_polya_tpm,embryo_intron_total_tpm)

  output_plot <- dataset_list[[i]] %>% select(- gene_id) %>% melt() %>%
    mutate(stage = factor(variable, levels = c('p3.5','p4.4','p4.8','p5.5','p6.5','p8','p10','p10_24'))) %>% select(-variable) %>%
    ggplot(aes(y= log2(value + 0.01) ,x= stage, fill=stage)) + scale_fill_brewer(palette="BuPu")  + geom_boxplot(notch=TRUE) +
    xlab(NULL) + ylab(NULL) + ggtitle(plot_titles[i]) +
    theme(plot.title = element_text(size=20, hjust = 0.5),axis.text.x = element_text(angle = 30, hjust = 1),axis.text = element_text(size = 14)) +
    scale_x_discrete(labels= c('0-9hs','11hs','12hs','13-14hs','16hs','20hs','24hs','24-60hs'))
  
  assign(plot_ids[i], output_plot)
  rm(output_plot)
}  

embryo_all_TPMcounts <-  ggarrange(p1,p2,p3,p4,nrow=2,ncol=2,common.legend = TRUE,legend = "none")
embryo_all_TPMcounts <- annotate_figure(embryo_all_TPMcounts, left = text_grob("Log2(TPM)", face = "bold", size = 18,rot = 90))
rm(p1,p2, p3,p4)
```

```{r}
#tiff("tpmCounts_allLibs.tiff", width = 10, height = 5.5, units = "in",res = 300,compression = "lzw")
embryo_all_TPMcounts
#dev.off()
```


## 3.3 Count reads - 5AZA- and DMSO-treated Emrbyos Total RNA-seq libraries

```{r}
embryo_5aza_dir <- "/drives/raid/AboobakerLab/data/bams/phaw/hisat/bams_05_2022/Aboobaker_embryo_5aza_tx"
embryo_5aza_bam <- grep(".bam$",list.files(embryo_5aza_dir,full.names = T),value=TRUE)


embryo_5aza_geneCounts <- featureCounts(embryo_5aza_bam,
                                         annot.ext =  "/drives/ssd1/manuel/phaw/2022_analysis/annotation/protein_coding_genes/transdecoder/unigenes/transcripts_cdhit_0.95.fa.transdecoder.genome_unigenes.gtf",
                                         isGTFAnnotationFile= TRUE,
                                         GTF.featureType = "transcript",
                                         GTF.attrType = "gene_id",
                                         useMetaFeatures= TRUE,
                                         allowMultiOverlap = FALSE,
                                         minOverlap = 2,
                                         strandSpecific = 2, #CHECK THIS PARAMETER
                                         isPairedEnd= TRUE,
                                         nthreads = 8)


embryo_5aza_exonCounts <- featureCounts(embryo_5aza_bam,
                                        annot.ext =  "/drives/ssd1/manuel/phaw/2022_analysis/annotation/protein_coding_genes/transdecoder/unigenes/transcripts_cdhit_0.95.fa.transdecoder.genome_unigenes.gtf",
                                        isGTFAnnotationFile= TRUE,
                                        GTF.featureType = "exon",
                                        GTF.attrType = "gene_id",
                                        useMetaFeatures= TRUE,
                                        allowMultiOverlap = FALSE,
                                        minOverlap = 2,
                                        strandSpecific = 2, #CHECK THIS PARAMETER
                                        isPairedEnd= TRUE,
                                        nthreads = 8)


embryo_5aza_intronCounts = embryo_5aza_geneCounts$counts - embryo_5aza_exonCounts$counts
embryo_5aza_intronCounts=pmax(embryo_5aza_intronCounts,0)
embryo_5aza_intronCounts = as.data.frame(embryo_5aza_intronCounts)
```

```{r}
embryo_5aza_exonCounts$counts %>% head()
```



```{r}
#Save count tables
write.table(embryo_exonCounts_polyA$counts,"embryo_exonCounts_polyA.txt",sep = "\t",quote = F,row.names = T)
write.table(embryo_exonCounts_total$counts,"embryo_exonCounts_total.txt",sep = "\t",quote = F,row.names = T)
```

